### PR TITLE
Add computed shard movement metric

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change adds a metric that would allow us to track amount of computed shard movements in a cluster

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java [hunk 1]
```diff
@@ -237,6 +237,7 @@ public DesiredBalanceStats getStats() {
             computationsExecuted.count(),
             computationsConverged.count(),
             desiredBalanceComputer.iterations.sum(),
+            computedShardMovements.sum(),
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:240]: We should expose average for the above 2 metrics

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java [hunk 2]
```diff
@@ -336,8 +336,8 @@ public DesiredBalance compute(
             assignments.put(shardAndAssignments.getKey(), ShardAssignment.ofAssignedShards(shardAndAssignments.getValue()));
         }
 
-        for (var ignored : routingNodes.unassigned().ignored()) {
-            var info = ignored.unassignedInfo();
+        for (var shard : routingNodes.unassigned().ignored()) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java:339]: Just to check I'm not missing anything here - we're just renaming `ignored -> shard` and extracting `unassigned ? 0 : 1` as `ignored`, not changing any behaviour right?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java [hunk 3]
```diff
@@ -30,4 +31,70 @@ public static boolean hasChanges(DesiredBalance a, DesiredBalance b) {
...
+            var oldAssignment = old.getAssignment(shardId);
+            var updatedAssignment = updated.getAssignment(shardId);
+            if (Objects.equals(oldAssignment, updatedAssignment) == false) {
+                movements += shardMovements(oldAssignment, updatedAssignment);
+            }
+        }
+        return movements;
+    }
+
+    public static int shardMovements(ShardAssignment old, ShardAssignment updated) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java:63]: Could we make this `private`? It's only exposed for tests AFAICT and we could test it through the one that compares at the `DesiredBalance` level instead.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java [hunk 4]
```diff
@@ -30,4 +31,70 @@ public static boolean hasChanges(DesiredBalance a, DesiredBalance b) {
...
+            }
+        }
+        if (movements > 0) {
+            // exclude newly assigned shards
+            movements -= Math.max(0, old.unassigned() - updated.unassigned());
+            // exclude new shard copies
+            movements -= Math.max(0, updated.total() - old.total());
+        }
+
+        return movements;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java:77]: Could we assert `movements >= 0`? I think that should always be true, although it's not totally obvious from the implementation.

**Review by DaveCTurner — COMMENTED** [review]
I think we discussed something like this a few weeks back but I forget exactly why we wanted it now. Could you remind me? Apart from my not understanding that, this looks ok to me (mod a few tiny comments below).

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java [hunk 5]
```diff
@@ -30,4 +31,70 @@ public static boolean hasChanges(DesiredBalance a, DesiredBalance b) {
...
+    }
+
+    public static String humanReadableDiff(DesiredBalance old, DesiredBalance updated) {
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java:80]: Correct again, moved from the actual allocator

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java (hunk 1 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:240]: Re above, what naming should be the best?
```
computation_iterations (already exists)
computation_iterations_avg
computed_shard_movements
computed_shard_movements_avg
```

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:240]: I'm not sure the average (over the life of the master) is all that useful. We usually care more about the average over a particular interval, which needs two API outputs taken at either end of the interval. Given that, I think we already expose enough metrics to compute the average: we divide the (difference in the) total by the (difference in the) number of computations executed, right?

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**